flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1021,203 +220,20 @@ public ArchivedExecutionConfig getArchivedExecutionConfig() {
      *
      * @return Termination future of this {@link ExecutionGraph}.
      */
-    public CompletableFuture<JobStatus> getTerminationFuture() {
-        return terminationFuture;
-    }
+    CompletableFuture<JobStatus> getTerminationFuture();
 
     @VisibleForTesting
-    public JobStatus waitUntilTerminal() throws InterruptedException {
-        try {
-            return terminationFuture.get();
-        } catch (ExecutionException e) {
-            // this should never happen
-            // it would be a bug, so we  don't expect this to be handled and throw
-            // an unchecked exception here
-            throw new RuntimeException(e);
-        }
-    }
-
-    // ------------------------------------------------------------------------
-    //  State Transitions
-    // ------------------------------------------------------------------------
-
-    public boolean transitionState(JobStatus current, JobStatus newState) {
-        return transitionState(current, newState, null);
-    }
-
-    private void transitionState(JobStatus newState, Throwable error) {
-        transitionState(state, newState, error);
-    }
+    JobStatus waitUntilTerminal() throws InterruptedException;
 
-    private boolean transitionState(JobStatus current, JobStatus newState, Throwable error) {
-        assertRunningInJobMasterMainThread();
-        // consistency check
-        if (current.isTerminalState()) {
-            String message = "Job is trying to leave terminal state " + current;
-            LOG.error(message);
-            throw new IllegalStateException(message);
-        }
+    boolean transitionState(JobStatus current, JobStatus newState);
 
-        // now do the actual state transition
-        if (state == current) {
-            state = newState;
-            LOG.info(
-                    "Job {} ({}) switched from state {} to {}.",
-                    getJobName(),
-                    getJobID(),
-                    current,
-                    newState,
-                    error);
+    void incrementRestarts();
 
-            stateTimestamps[newState.ordinal()] = System.currentTimeMillis();
-            notifyJobStatusChange(newState, error);
-            return true;
-        } else {
-            return false;
-        }
-    }
+    void initFailureCause(Throwable t);
 
-    public void incrementRestarts() {
-        numberOfRestartsCounter.inc();
-    }
+    void vertexFinished();
```

Comment:
If we made sure that the Execution* classes all work against the DefaultExecutionGraph then we should be able to remove this and other package-private methods.

Reply:
Yes, I considered this as well. But the problem is that this introduces a dependency on DefaultExecutionGraph again.
For example here: https://github.com/apache/flink/blob/master/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java#L672 , we would require a DefaultExecutionGraph in a test mock, voiding the benefits of this effort.

Reply:
That is only necessary because Executing works directly against ExecutionJobVertices/ExecutionVertices.
If there would be 2 small interfaces exposing the vertices the the JobVertex and deploy/markFailed for the ExecutionVertex, then you wouldn't need to mock these vertices in the test and would no longer require package-private APIs.

As it stands you're exposing a whole bunch of package-private methods to the outside which is just not acceptable.

Reply:
Sorry to ask again, but I don't seem to understand your suggestion here (and I'm having the feeling I'm overlooking something obvious)

None of the package-private methods are used in the context of the ExecutionGraph interface. The only method I'm using in the `Executing.deploy()` method is `getVerticesTopologically()`, which was public before.
deploy/markFailed are private methods of `ExecutionVertex`, which I don't expose in this PR.

Introducing an interface for `ExecutionJobVertex` (with a `.getDeployableTaskVertices()`) and `ExecutionVertex` (with a `.deploy()` and `.markFailed()` ) would indeed simplify this test, but seems out of scope of this change.

Reply:
Yes I didn't explain it very well, and I don't think my suggestion (according to the phrasing) as is would work particularly well.

The original issue we wanted to solve with this PR is that mocking an EG is annoying and error-prone, so we introduced an interface to remedy that.
However, because all the Execution* classes are pretty much one intermingled mess, and the Executing state works against the Execution(Job)Vertex, you ended up still having a dependency on the package-private APIs, which you also exposed via the interface.

You them introduced an interface for the EG containing the internal APIs; this allows you to mock the internal API of the EG that the vertex requires. While it works it doesn't really address the core issue of not having a clear API. Despite only wanting to assert that the state deploys/cancels some vertices we end up having to mock EG-internal APIs and all that jazz.

Alternatively, and this is what I was more going for, we could go with a similar approach like what you did for the EG for the remaining Execution* classes. In that model you'd have to mock a lot less methods, and can entirely ignore how these classes internally communicate.
I would think that this wouldn't be soooo much overhead, in particular because we need a lot less methods from these classes; so we can probably get by with a tiny extension of the Access* interfaces.

Reply:
To add on to that, in the long-term interfaces for all Execution* classes would be pretty sick because we could re-implement the EG as we see fit, whereas an interface between the EG and vertices is probably not that beneficial.

Reply:
Thanks a lot for your explanation. Now I understand what you mean.

How would you feel about merging this PR as-is, because it already improves the situation a lot, and we address the introduction of interfaces for the remaining Execution* classes in a follow up?